caikit_nlp/toolkit/text_generation/model_run_utils.py

```diff
@@ -237,7 +237,9 @@ def generate_text_func(
         generate_ids[0, -1] == tokenizer.eos_token_id
     ):
         finish_reason = "EOS_TOKEN"
-    elif generate_ids.size(1) - 1 == max_new_tokens:
+    elif (generate_ids.size(1) - 1 == max_new_tokens) or (
+        generate_ids.size(1) - inputs["input_ids"].size(1) == max_new_tokens
```

Comment:
hmm. technically input will only be in output for `causal-lm` type of models, so there can be some side-effects in this calculations  🤔  May be we would need to check if input is actually in output and then calculate max_new_tokens for that accordingly

Reply:
@gkumbhat  I agree. I gave some thought to it and I ended up with the logic in the latest push. The idea is the following:

1. Check if the last token is `eos_token`, if so, `finish_reason = "EOS_TOKEN"`
2. Check if the stop sequence is the the generated tokens, if so, `finish_reason = "STOP_SEQUENCE"`
3. If none of the above conditions are true, then `finish_reason = "MAX_TOKENS"`

My reasoning for always returning `MAX_TOKENS` comes from looking at [`caikit.interfaces.nlp.data_model.text_generation.FinishReason`]( https://github.com/caikit/caikit/blob/4d01c9b4caded5492ca5772875f0d8778143a657/caikit/interfaces/nlp/data_model/text_generation.py#L35-L44):

Ia added some comments to each enum value
```python
@dataobject(package=NLP_PACKAGE)
class FinishReason(Enum):
    NOT_FINISHED = 0 # should only be returned by streaming implementations
    MAX_TOKENS = 1  
    EOS_TOKEN = 2  # matches rule #1 above
    CANCELLED = 3  # should not be set by generate_text_func
    TIME_LIMIT = 4 # should not be set by generate_text_func
    STOP_SEQUENCE = 5 # matches rule #2 above
    TOKEN_LIMIT = 6 # not sure about this
    ERROR = 7  # fairly generic, 
```

I'm unsure about `TOKEN_LIMIT` (how is this different from `MAX_TOKENS`?), but I think none of the other enum values apply to `generate_text_func` .

Reply:
`TOKEN_LIMIT`refers to the maximum number of tokens limit defined by the model whereas the `MAX_TOKENS` refers to the maximum number defined by the user. So one can reach `TOKEN_LIMIT` before `MAX_TOKENS`

Reply:
@gkumbhat Ok, so that means that also need to check whether we reached the token limit for the model, although I'm not sure how we can get that information. Can we open an issue for adding support for `TOKEN_LIMIT` so that we can work on that on another PR?

Reply:
Created https://github.com/caikit/caikit-nlp/issues/253